### DISCUSSION TITLE
Contributions guide/checklist

### DISCUSSION THREAD

**Discussion by almirsarajcic** [3Stones-io/omedis#244] [General]
I'd like us to have some kind of checklist that we go through whenever we work on something.
We already have some kind of process, but it's kinda implicit or discussed privately.
So let's formalize it and be explicit about it so the onboarding gets easier.

I guess when we settle on it, we can also create a CONTRIBUTING.md file in the repo.

So far, what I noticed we do is:
1. check the issue you're starting to work on to see whether it provides sufficient details and ask questions
2. divide the issue into multiple smaller chunks if there's a need
3. add the issue to the project (https://github.com/users/wintermeyer/projects/1) and move it to the "In progress" column
4. create the PR, request a review, and move the issue to the "In review" column

Details:
- resource creation/update
  - make sure you run the `mix ash_postgres.generate_migrations`
  - check the migration and update until you're satisfied (it looks as if you were writing in by hand)
  - resource snapshots need to be commited
  - add seeds to `demo_seeds.exs` file if there's a need
- PR creation
  - make sure code checks run successfully locally before pushing the code (`mix check_code`)
  - link to an issue by writing "Fixes #<issue_number>"
  - provide screen recording or screenshot if there are changes in the UI
  - before requesting a review, make sure your PR has a green checkmark (GitHub Actions workflow runs successfully)

Let's hear more ideas.

**Comment by wintermeyer** [comment]
Add a screenshot or a screencast when ever possible. With the initial issue and the PR. Of course it takes time for the person who is working on the issue but it saves so much time for everybody who looks at the issue/PR the first time.

**Comment by almirsarajcic** [comment]
Keep the PRs short and stay on topic.
Create a new PR if you want to include some refactoring or other unrelated changes.

**Comment by almirsarajcic** [comment]
Include translations for any messages and UI you add.
Thanks, @amos-kibet. I completely forgot to add translations to the messages I was adding.

**Comment by amos-kibet** [comment]
Check existing code for related features, or ask the developer who implemented the feature, when faced with challenges in implementing a feature. This way may be faster, rather than trying to figure it out which may take longer.
Take this error for example:
```
** (ArgumentError) `nil` is not a Spark DSL module.
stacktrace:
       nil.persisted(:data_layer)
       (spark 2.2.35) lib/spark/dsl/extension.ex:138: Spark.Dsl.Extension.persisted!/3
       (ash 3.4.36) lib/ash/filter/filter.ex:3708: Ash.Filter.do_hydrate_refs/2
       (ash 3.4.36) lib/ash/filter/filter.ex:3571: Ash.Filter.do_hydrate_refs/2
       (omedis 0.1.0) lib/omedis/accounts/access_filter.ex:5: Omedis.Accounts.AccessFilter.try_eval/2
       (omedis 0.1.0) lib/omedis/accounts/access_filter.ex:5: Omedis.Accounts.AccessFilter.try_strict_check/3
```
Since this error is related to policies, it may be faster to ask @almirsarajcic, or check existing policies-related code.

**Reply by fradarr** [reply]
Definitely the way to go, but generaly how do we know who wrote the code so we can adress the right person?

**Reply by amos-kibet** [reply]
When you open the file on GitHub, you can click on the **blame** button, and it will show you the person who wrote the code.

https://github.com/user-attachments/assets/6e4812c2-61ad-49af-98ad-2d3e3442454e

Alternatively, you can use VS Code's GitLens extension (https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens), where you just hover on the code, and it shows you the person who wrote that code.

https://github.com/user-attachments/assets/692525eb-0075-4d84-8ca9-34712ae94f93
